### ISSUE TITLE
Using multi GPU

### ISSUE THREAD

**Issue by Armin234** [GPUOpen-LibrariesAndSDKs/AMF#132]
Hi,
do i have to take care of anything if i want to use more than one GPU?
I'm working one a PC with one Radeon RX Vega 64 and one Radeon Vega Frontier Edition.
The final system will have four WX9100.

Regards

**Comment by MikhailAMD** [comment]
Assumption is that AMF is a small part of a larger application and the application actively uses GPU and should be able to select a GPU for all GPU activities. So in multi-GPU case it makes sense that application selects GPU and passes it to AMF context. AMF has several examples for GPU enumeration: DeviceDX9.cpp, DeviceDX11.cpp, DeviceOpenCL.cpp, DeviceVulkan.cpp, DeviceOpneGL.cpp. The default case when an app passes NULL to AMFContext::Init<>() functions will work but AMF will select a default GPU.

**Comment by Armin234** [comment]
f I create two contexts and let each GPU perform its own task, then it works well.
I have problems when GPU 1 encodes a Image and the result should be decoded by both GPU's for having the decoded Image in the memory of both GPU's.
How can I pass the result of QueryOutput of the GPU1 with Submitinput of the GPU2, do I have to allocate a new memory and copy the content?

**Comment by MikhailAMD** [comment]
If you want to copy a compressed video frame, it is returned in system memory. You would have to allocate a AMFBuffer on the second context with AMF_MEMORY_HOST, call GetNative() from the first buffer and memcpy the data. Compressed data are relatively small so it will be reasonably fast.
If you want to transfer uncompressed video frame from one GPU to another there is DirectGMA API but it is available on PRO GPU/drivers only. Check https://gpuopen.com/gaming-product/direct-gma/ and https://github.com/GPUOpen-LibrariesAndSDKs/DirectGMA_P2P/tree/master/GPUtoGPU_DX11
Since DirectGMA requires special allocation you may need to GPGPU copy from decoder output to DirectGMA texture (ID3D11DeviceContext::CopyResource).  
For non-pro drivers you would have to use Map/memcpy/Unmap sequence.

**Comment by Armin234** [comment]
So I tried it too, but the decoded images are then of unusable quality. I use the SetPTS to number the images, because of the race condition of the threads, the Image to decode are not sequentiell.

**Comment by MikhailAMD** [comment]
1. Tried what?
2. Frames and PTS can be reordered by decoder. You should set a private property on submission buffer and read it back from output surface. If reordering happens the property will come reordered . Or you can disable PTS reordering (frames will be reordered anyway): decoder->SetProperty(AMF_TIMESTAMP_MODE, AMF_TS_DECODE).
3. Decoder is a single resource. If you run two decoders you may consider pace the submission to avoid long waits. 
4. Quality of encoding and decoding does not depend on order of submission. Please clarify what "unusable quality" means.
5. If you want low latency decoding (one -in - one-out) and sure that your input stream doesn't need reordering (no B-frames) you may want to enable decoder's low latency mode: decoder->SetPropery(AMF_VIDEO_DECODER_REORDER_MODE,  AMF_VIDEO_DECODER_MODE_LOW_LATENCY);

**Comment by Armin234** [comment]
![quality](https://user-images.githubusercontent.com/37782776/45318793-3fee9400-b53e-11e8-95bb-7b15ee24d3a7.PNG)

I attached a screenshot that shows the decoded Images and the Original.
With decoder->SetProperty(AMF_TIMESTAMP_MODE, AMF_TS_DECODE) and private Property
i have Always the same problems. I'm already using  low latency decoding.

**Comment by Armin234** [comment]
Yes, i encode one Image with one GPU and decode the encoded Frame on two GPUs in parallel.
You can download my Files on https://we.tl/t-zqsOnnGqSF

I don't use extrabuffer and i use DX11

I will try the AMFdump

**Comment by Armin234** [comment]
I inserted the following Code, but the Encoder  didn't dump 


		void* pInterf = NULL;
		amf::AMFDumpPtr m_pInterf;

		if (encoder[uu]->QueryInterface(amf::AMFDump::IID(), &pInterf) == AMF_OK)
		{
			m_pInterf = static_cast<amf::AMFDump*>(pInterf);
			m_pInterf->EnableOutputDump(true);

			m_pInterf->SetDumpBasePath(L"c:\\tmp");
		}

**Comment by MikhailAMD** [comment]
You have to initialize dump before Init() call.

**Comment by Armin234** [comment]
There must be another bug in my code, because the Dump still doesn't work

		res = g_AMFFactory.GetFactory()->CreateComponent(context[uu], USE_ENCODER, &encoder[uu]);
		res = encoder[uu]->SetProperty(AMF_VIDEO_ENCODER_PROFILE, AMF_VIDEO_ENCODER_PROFILE_HIGH);
		res = encoder[uu]->SetProperty(AMF_VIDEO_ENCODER_PROFILE_LEVEL, 51);
		res = encoder[uu]->SetProperty(AMF_VIDEO_ENCODER_B_PIC_PATTERN, 0);

		void* pInterf = NULL;

		if (encoder[uu]->QueryInterface(amf::AMFDump::IID(), &pInterf) == AMF_OK)
		{
			m_pInterf[uu] = static_cast<amf::AMFDump*>(pInterf);
			
			if (uu == 0)
			{
				m_pInterf[uu]->SetDumpBasePath(L"c:\\temp\\0\\");
			}
			else
			{
				m_pInterf[uu]->SetDumpBasePath(L"c:\\temp\\1\\");
			}

			m_pInterf[uu]->EnableOutputDump(true);
		}
		res = encoder[uu]->Init(USED_AMF_FORMAT, lW, lH);

**Comment by Armin234** [comment]
Okay, without  m_pInterf[uu]->SetDumpBasePath(…) the dumping works. see the file in attachment with the same errors

[output_2456x3684.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/2370608/output_2456x3684.zip)

**Comment by MikhailAMD** [comment]
I am confused:
1. I've got an email with a link to the dumped file. So does dump work for you or not?
2. You may have memory leak: when you get dump interface QueryInterface() will increase reference count on component. if m_pInterf[uu] is a smart pointer when assigned it will increase reference count again.
3. Looks like the dump file output_2456x3684.zip has input corruption. H264 stream is compliant. How do you produce input?
4. How do you do color conversions?

**Comment by Armin234** [comment]
Excuse me, I messed things up with my different attempts.

1. I must call EnableOutputDump in the same thread with the QueryOutput call, then Dump is working.
2. with PlaybackHW  the dumps are looking good, and my saved h264 too.
3. the output off the decoder is corrupted ( L0_2.Bmp …. )

4. I'm not making color conversions, I'm using AMF_SURFACE_RGBA

[Armin.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/2371985/Armin.zip)
[SimpleEncoder.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/2371992/SimpleEncoder.zip)

**Comment by MikhailAMD** [comment]
It seem in PollingThreadEncB::Run() you write across GPUs ready buffers and flags and mixing things up: 
GPU0 writes to GPU0 and GPU1 from thread 1
GPU1 writes to GPU1 and GPU0 from thread 2
So frames from H264 streams can be mixed up
			amf::AMFBufferPtr buffer(lDataComp);
			mCompressedData[m_GPU][uu] = buffer;

			for (int jj = 0; jj < ANZGPU; jj++)
			{
				if(jj != m_GPU)

**Comment by Armin234** [comment]
Yes, Thread 1 (handles GPU0, mGPU == 0) and elaborates Frame with Property-Tag uu= 0 or 2, after the success of QueryOutput it saves the pointer in mCompressedData[0][uu] = Buffer and make a copy buffer2 for the other GPU  mCompressedData[1][uu]=buffer2

Thread 2 (handles GPU1, mGPU == 1) and elaborates Frame with Property-Tag uu= 1 or 3

Frame 0 should be encoded by GPU0 and then decoded by GPU0 and by GPU1 
Frame 2 should be encoded by GPU0 and then decoded by GPU0 and by GPU1 
Frame 1 should be encoded by GPU1 and then decoded by GPU0 and by GPU1
Frame 3 should be encoded by GPU1 and then decoded by GPU0 and by GPU1

the Decoding is done by  thread PollingThreadDecA for every GPU

I tagged the frames with the private property "TAG" so the mixing isn't a Problem for me.

But why the Output is corrupted ?

**Comment by Armin234** [comment]
Okay, now i understand the problem, i assumed that every frame is I-Frame with full frame info.
Since i must handle single Images, it should be fine when every Frame is a I-Frame, is this possible ?

**Comment by MikhailAMD** [comment]
Yes, you can force frame type if you set property on **every input encoder frame** : surface->SetProperty(AMF_VIDEO_ENCODER_FORCE_PICTURE_TYPE,  AMF_VIDEO_ENCODER_PICTURE_TYPE_I) or IDR.
But don't expect any meanful bitrate for output.

**Comment by Armin234** [comment]
Now it's look good, thanks !!!!
